circuit/types/field/src/equal.rs

```diff
@@ -32,24 +33,41 @@ impl<E: Environment> Equal<Self> for Field<E> {
     /// This method constructs a boolean that indicates if
     /// `self` and `other ` are *not* equal to each other.
     ///
-    /// This method costs 3 constraints.
+    /// This method costs 2 constraints.
     ///
     fn is_not_equal(&self, other: &Self) -> Self::Output {
+        // In all cases, the witness ("ejected") value is calculated from the ejected values.
+        let is_neq_ejected = self.eject_value() != other.eject_value();
+
         match (self.is_constant(), other.is_constant()) {
-            (true, true) => witness!(|self, other| self != other),
+            // If both operands are constant, the result is also constant.
+            (true, true) => Boolean::new(Mode::Constant, is_neq_ejected),
+
+            // Otherwise, we introduce a private field variable is_neq for the result,
+            // along with an auxiliary variable multiplier for the inverse of the difference between the operands,
+            // and enforce the following constraints:
+            //   (self - other) (multiplier) = (is_neq)
+            //   (self - other) (1 - is_neq) = (0)
+            // These constraints imply that is_neq is boolean, i.e. either 0 or 1;
+            // so we avoid creating is_neq as a Boolean, which would generate an unneeded boolean constraint.
+            // See the comments just after the code for more detailed explanations.
             _ => {
-                // Compute a boolean that is `true` if `this` and `that` are not equivalent.
-                let is_neq: Boolean<E> = witness!(|self, other| self != other);
-
-                // Assign the expected multiplier.
-                let multiplier: Field<E> = witness!(|self, other| {
-                    match (self - other).inverse() {
+                let is_neq = Boolean::from_variable(E::new_variable(Mode::Private, match is_neq_ejected {
```

Comment:
Can you convince me this is safe?

Without a booleanity check, a malicious actor can allocate `is_neq` with any field value. (As we no longer enforce 0/1 field).

Reply:
Yes, good question. Here is a proof that the remaining constraints
```
1.   (self - other) (multiplier) = (is_neq)
2.   (self - other) (1 - is_neq) = (0)
```
imply that `is_neq` is boolean. There are two cases: either `self = other` or `self != other`. If they are equal, constraint 1 implies `is_neq = 0`. If they are not equal, constraint 2 implies `1 - is_neq = 0`, i.e. `is_neq = 1`. So `is_neq` is either 0 or 1. If a malicious actor tries to construct a solution to the constraints with `is_neq` neither 0 nor 1, the system of two constraints won't be satisfied.

Another way to come to the same conclusion is to prove the correctness of this circuit, i.e. that the two constraints are equivalent to the specification of the operation, which is
```
is_neq = [IF self = other THEN 0 ELSE 1]
```
and which shows that `is_neq` can only be 0 or 1. (I can spell out this correctness proof if you want.)

Reply:
This is similar to what we do in boolean operations like [and](https://github.com/AleoHQ/snarkVM/blob/9cd5a323c611b65d2042282d0398d9eec88303ff/circuit/types/boolean/src/and.rs#L83-L96): note the comment there about `output` being guaranteed boolean by the constraints. (Side note: the boolean operations require preconditions that the operands are boolean; field (in)equality has no preconditions, works on any field values `self` and `other`.)

Reply:
Just in case you are curious, the ACL2 proof of the optimized circuit in PFCS form is [here](https://github.com/AleoHQ/aleo-acl2/blob/master/circuits/pfcs/field-neq-opt.lisp).